includes/multi-currency/class-multi-currency.php

```diff
@@ -187,21 +203,66 @@ public function enqueue_scripts() {
 	}
 
 	/**
-	 * Gets the mock available.
+	 * Wipes the cached currency data option, forcing to re-fetch the data from WPCOM.
 	 *
-	 * @return array Array of currencies.
+	 * @return void
 	 */
-	public function get_mock_currencies() {
+	public function clear_cache() {
+		delete_option( self::CURRENCY_CACHE_OPTION );
+	}
+
+	/**
+	 * Gets and caches the data for the currency rates from the server.
+	 * Will be returned as an array with three keys, 'currencies' (the currencies), 'expires' (the expiry time)
+	 * and 'updated' (when this data was fetched from the API).
+	 *
+	 * @return ?array
+	 */
+	public function get_cached_currencies() {
+		if ( ! $this->payments_api_client->is_server_connected() ) {
+			return null;
+		}
+
+		$cache_data = $this->read_currencies_from_cache();
+
+		// If the option contains the error value, return false early and do not attempt another API call.
+		if ( isset( $cache_data['currencies'] ) && self::CURRENCY_RETRIEVAL_ERROR === $cache_data['currencies'] ) {
+			return null;
+		}
+
+		// If an array of currencies was returned from the cache, return it here.
+		if ( null !== $cache_data ) {
+			return $cache_data;
+		}
+
+		// If the cache was expired or something went wrong, make a call to the server to get the
+		// currency data.
+		try {
+			$currency_from = get_woocommerce_currency();
+			$currencies_to = get_woocommerce_currencies();
```

Comment:
Hm, it's odd that the request is timing out. In my case it's returning within about ~1200ms consistently. Is that using your local server or your sandbox?

The idea we had was that we'd make the request to the server once on first load of the multi-currency and then cache for 6 hours afterwards, although sending a GET request with so many currencies appended to the query body does make me slightly uncomfortable.

The only other alternative I can think of is that we change the server request so that if no `currencies_to` is supplied, it'll fetch all the currencies. If we did that, we'd need to have a list of the currencies supported by WooCommerce (the list currently returned by `get_woocommerce_currencies()` stored on the server. 

Do you see an issue with us doing that? My concern would be that the list would get updated in WC and we'd have to make sure we kept the server list up to date.

Reply:
> Instead of maintaining a list of currencies for when there's no currencies_to param, we could pull it from our exchange rate provider.

We do maintain a list of these already: https://opengrok.a8c.com/source/xref/trunk/wp-content/admin-plugins/wpcom-billing/class-store-exchange-rate.php?r=204367&mo=58&fi=3#3

I could update this code to add a way of fetching that list for use in the server. My only concern really is that there may be currencies which are supported by WooCommerce, but not listed in the OpenExchange rate that we may want to support (by allowing the merchant to sell in that currency setting a manual rate).

I supposer I could get around that by getting the OpenExchange supported currency rates from the server, but on the client using `get_woocommerce_currencies()` and then looping over them, initialising them with an automatic rate (if one is available) or 0 if it is not. 

Glad you were able to figure out why it was timing out! Will be pushing a commit to move towards using the `init` filter now. 👍